caas/kubernetes/provider/application/application.go

```diff
@@ -392,6 +388,158 @@ func (a *app) Exists() (caas.DeploymentState, error) {
 	return state, nil
 }
 
+func headlessServiceName(appName string) string {
+	return fmt.Sprintf("%s-endpoints", appName)
+}
+
+func (a *app) configureHeadlessService(name string, annotation annotations.Annotation) error {
+	svc := resources.NewService(headlessServiceName(name), a.namespace, &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels: a.labels(),
+			Annotations: annotation.
+				Add("service.alpha.kubernetes.io/tolerate-unready-endpoints", "true"),
+		},
+		Spec: corev1.ServiceSpec{
+			Selector:                 a.labels(),
+			Type:                     corev1.ServiceTypeClusterIP,
+			ClusterIP:                "None",
+			PublishNotReadyAddresses: true,
+		},
+	})
+	return svc.Apply(context.Background(), a.client)
+}
+
+// configureDefaultService configures the default service for the application.
+// It's only configured once when the application was deployed in the first time.
+func (a *app) configureDefaultService(annotation annotations.Annotation) (err error) {
+	svc := resources.NewService(a.name, a.namespace, &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:      a.labels(),
+			Annotations: annotation,
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: a.labels(),
+			Type:     corev1.ServiceTypeClusterIP,
+		},
+	})
+	if err = svc.Get(context.Background(), a.client); errors.IsNotFound(err) {
+		return svc.Apply(context.Background(), a.client)
+	}
+	return errors.Trace(err)
+}
+
+// UpdateService updates the default service with specific service type and port mappings.
+func (a *app) UpdateService(param caas.ServiceParam) error {
```

Comment:
This one is more like a place holder and it won't be called in the new caasfirewaller worker until we have juju expose modelled properly.
It could be ignored for now.